File: src/main/scala/com/pattern/FactoryMethod.scala
package com.pattern

/*
 * <p> Creation Pattern </p>
 * Common Interface to Object Creation
 * Hide complex object creation logic
 * Encapsulate object construction
 * 
 * Pros 
 * use trait name for factory
 * use of apply method to construct object
 * 
 */

object FactoryMethod {
  def main(args: Array[String]) {
    println(Animal("cat"))
  }
}

trait Animal 
class Dog extends Animal
class Cat extends Animal

object Animal {
  def apply(kind: String) = kind match {
    case "cat" => new Cat
    case "Dog" => new Dog
  }
}

File: src/main/java/com/pattern/FactoryMethod.java
package com.pattern;

/**
 * @author sumit
 *	Cons 
 *	Cant user Animal interface to define factory method
 *	Need Extra class to define factory method 
 */
public class FactoryMethod {
	public static void main(String[] args) {

	}
}

interface Animal {}
class Dog implements Animal {}
class Cat implements Animal {}

class AnimalFactory {
	public static Animal newAnimal(String kind) {
		if("cat".equals(kind)) return new Cat();
		if("dog".equals(kind)) return new Dog();
		throw new IllegalArgumentException();
	}
}
